modules/internal/DefaultAssociative.chpl

```diff
@@ -566,14 +566,15 @@ module DefaultAssociative {
 
     // ref version
     proc dsiAccess(idx : idxType, haveLock = false) ref {
-      const shouldLock = dom.parSafe && !haveLock;
+      const arrOwnsDom = dom._arrs.length == 1;
+      const shouldLock = dom.parSafe && !haveLock && arrOwnsDom;
```

Comment:
Re my example, I meant it to be the long form of the simple forall loop you've been using (repeated above). I just didn't write it right. Here's another attempt:
``` chapel
forall key in collection {
  if D.member(key) then
    A[key] += 1;
  else {
   D += key;
   A[D] = 1;
  }
}
```

My reasoning is this:
 * I think the modifies-domain stuff is a convenience/shorthand and so shouldn't change memory consistency / data races from how you'd write it without using the shortcut
 * I don't see how to write it in a race-free way without using the short form

Therefore I'm questioning:
 a) whether we want the special 1:1 behavior at all
 b) whether it makes sense for the existence of the 1:1 behavior to impact our parallel semantics. That is, if the "long form" of the same loop obviously wouldn't be thread safe, why should we go to effort in the implementation to make the short form be so?

I think I'd rather expect the user to use a more general mechanism (it could be - these statements happen under the domain's lock; or require them to use a separate lock; or ideally the `atomic` statement).